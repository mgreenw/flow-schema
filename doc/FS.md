# API Documentation

_NOTE_: These docs are still a work in progress.

```js
import FS from 'flow-schema';
```

## Schema

A `Schema<T>` is the main object class in `flow-schema`. A `Schema<T>` has the following methods/properties:

- `validate(value: mixed): T`
- `get optional(): Schema<T | void>`
- `get nullable(): Schema<T | null>`
- `get maybe(): Schema<T | null | void>`
- `get nonMaybe(): Schema<$NonMaybeType<T>>`

## Base Type Schemas

_Here are the base type schemas and their Flow equivalent._

| Type    | Schema     |
|---------|------------|
| void    | FS.void    |
| null    | FS.null    |
| string  | FS.string  |
| boolean | FS.boolean |
| number  | FS.number  |
| Date    | FS.Date    |
| any     | FS.any     |
| mixed   | FS.mixed   |

## Type Schema Constructors (by example)

_Here are the more powerful "schema constructors" that, when called, these each return a `Schema` that validates to the listed Flow equivalent._
s

| Type                  | Constructor                               |
|-----------------------|-------------------------------------------|
| {&#124; a: A, b: B &#124;}      | FS.Object({ a: Schema<A>, b: Schema<B> }) |
| Array<T>              | FS.Array(Schema<T>)                       |
| { [string]: T }       | FS.ObjectMap(Schema<T>)                   |
| A & B                 | FS.intersection(Schema<A>, Schema<B>)     |
| A | B                 | FS.union(Schema<A>, Schema<B>)            |
| $Keys<{ a: A, b: B }> | FS.$Keys({ a: Schema, b: Schema })        |
| $ReadOnly<T>          | FS.$ReadOnly(Schema<T>)                   |
| $ReadOnlyArray<T>     | FS.$ReadOnlyArray(Schema<T>)              |
| [A, B]                | FS.tuple(Schema<A>, Schema<B>)            |
| T                     | FS.literal<T>(Schema<T>)                  |