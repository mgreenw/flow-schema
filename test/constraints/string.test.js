// @flow

import FS from '../..';

describe('String Constraints', () => {
  it('min', () => {
    expect(() => FS.string.min(10).validate('123456789')).toThrowErrorMatchingSnapshot();
    expect(() => FS.string.min(3.5).validate('123')).toThrowErrorMatchingSnapshot();
    expect(FS.string.min(3).validate('1234')).toBe('1234');

    // Cannot have a negative length for the min
    expect(() => FS.string.min(-10)).toThrowErrorMatchingSnapshot();
  });

  it('max', () => {
    expect(() => FS.string.max(10).validate('12345678910')).toThrowErrorMatchingSnapshot();
    expect(() => FS.string.max(8).validate('123456789101112')).toThrowErrorMatchingSnapshot();
    expect(FS.string.max(3).validate('12')).toBe('12');
    expect(FS.string.max(0).validate('')).toBe('');

    // Cannot have a negative length for the max
    expect(() => FS.string.max(-1)).toThrowErrorMatchingSnapshot();
  });

  it('length', () => {
    expect(() => FS.string.length(10).validate('123456789')).toThrowErrorMatchingSnapshot();
    expect(() => FS.string.length(8).validate('123456789')).toThrowErrorMatchingSnapshot();
    expect(FS.string.length(0).validate('')).toBe('');
    expect(FS.string.length(3).validate('123')).toBe('123');

    // Cannot have a negative length for the length
    expect(() => FS.string.length(-4)).toThrowErrorMatchingSnapshot();
  });

  it('uppercase', () => {
    expect(() => FS.string.uppercase.validate('not ALL UPPERCASE')).toThrowErrorMatchingSnapshot();
    expect(() => FS.string.uppercase.validate('abc123')).toThrowErrorMatchingSnapshot();
    expect(FS.string.uppercase.validate('')).toBe('');
    expect(FS.string.uppercase.validate('123')).toBe('123');
    const result = FS.string.uppercase.validate('IS ALL UPPERCASE');
    (result: string);
    expect(result).toBe('IS ALL UPPERCASE');
  });

  it('lowercase', () => {
    expect(() => FS.string.lowercase.validate('not ALL LOWERCASE')).toThrowErrorMatchingSnapshot();
    expect(() => FS.string.lowercase.validate('ABC123')).toThrowErrorMatchingSnapshot();
    expect(FS.string.lowercase.validate('')).toBe('');
    expect(FS.string.lowercase.validate('123')).toBe('123');
    const result = FS.string.lowercase.validate('is all lowercase');
    (result: string);
    expect(result).toBe('is all lowercase');
  });
});
