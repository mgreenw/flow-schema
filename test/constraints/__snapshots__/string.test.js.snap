// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`String Constraints length 1`] = `
"ValidationError

Expected: string /* length === 10 */
Received: \\"123456789\\" (String)

Violated Constraints (1):
-  length === 10
"
`;

exports[`String Constraints length 2`] = `
"ValidationError

Expected: string /* length === 8 */
Received: \\"123456789\\" (String)

Violated Constraints (1):
-  length === 8
"
`;

exports[`String Constraints length 3`] = `"String length constraint must be positive (received -4)"`;

exports[`String Constraints lowercase 1`] = `
"ValidationError

Expected: string /* lowercase */
Received: \\"not ALL LOWERCASE\\" (String)

Violated Constraints (1):
-  lowercase
"
`;

exports[`String Constraints lowercase 2`] = `
"ValidationError

Expected: string /* lowercase */
Received: \\"ABC123\\" (String)

Violated Constraints (1):
-  lowercase
"
`;

exports[`String Constraints max 1`] = `
"ValidationError

Expected: string /* <= 10 */
Received: \\"12345678910\\" (String)

Violated Constraints (1):
-  <= 10
"
`;

exports[`String Constraints max 2`] = `
"ValidationError

Expected: string /* <= 8 */
Received: \\"123456789101112\\" (String)

Violated Constraints (1):
-  <= 8
"
`;

exports[`String Constraints max 3`] = `"Maximum string length constraint must be positive (received -1)"`;

exports[`String Constraints min 1`] = `
"ValidationError

Expected: string /* length >= 10 */
Received: \\"123456789\\" (String)

Violated Constraints (1):
-  length >= 10
"
`;

exports[`String Constraints min 2`] = `
"ValidationError

Expected: string /* length >= 3.5 */
Received: \\"123\\" (String)

Violated Constraints (1):
-  length >= 3.5
"
`;

exports[`String Constraints min 3`] = `"Minimum string length constraint must be positive (received -10)"`;

exports[`String Constraints uppercase 1`] = `
"ValidationError

Expected: string /* uppercase */
Received: \\"not ALL UPPERCASE\\" (String)

Violated Constraints (1):
-  uppercase
"
`;

exports[`String Constraints uppercase 2`] = `
"ValidationError

Expected: string /* uppercase */
Received: \\"abc123\\" (String)

Violated Constraints (1):
-  uppercase
"
`;
